airflow/operators/python.py

```diff
@@ -145,6 +147,141 @@ def execute_callable(self):
         return self.python_callable(*self.op_args, **self.op_kwargs)
 
 
+class _PythonFunctionalOperator(BaseOperator):
+    """
+    Wraps a Python callable and captures args/kwargs when called for execution.
+
+    :param python_callable: A reference to an object that is callable
+    :type python_callable: python callable
+    :param multiple_outputs: if set, function return value will be
+        unrolled to multiple XCom values. List/Tuples will unroll to xcom values
+        with index as key. Dict will unroll to xcom values with keys as keys.
+        Defaults to False.
+    :type multiple_outputs: bool
+    """
+
+    template_fields = ('_op_args', '_op_kwargs')
+    ui_color = '#ffefeb'
+
+    # since we won't mutate the arguments, we should just do the shallow copy
+    # there are some cases we can't deepcopy the objects(e.g protobuf).
+    shallow_copy_attrs = ('python_callable',)
+
+    @apply_defaults
+    def __init__(
+        self,
+        python_callable: Callable,
+        multiple_outputs: bool = False,
+        *args,
+        **kwargs
+    ) -> None:
+        # Check if we need to generate a new task_id
+        task_id = kwargs.get('task_id', None)
+        dag = kwargs.get('dag', None) or DagContext.get_current_dag()
+        if task_id and dag and task_id in dag.task_ids:
+            prefix = task_id.rsplit("__", 1)[0]
+            task_id = sorted(
+                filter(lambda x: x.startswith(prefix), dag.task_ids),
+                reverse=True
+            )[0]
+            num = int(task_id[-1] if '__' in task_id else '0') + 1
+            kwargs['task_id'] = f'{prefix}__{num}'
+
+        if not kwargs.get('do_xcom_push', True) and not multiple_outputs:
+            raise AirflowException('@task needs to have either do_xcom_push=True or '
+                                   'multiple_outputs=True.')
+        if not callable(python_callable):
+            raise AirflowException('`python_callable` param must be callable')
+        self._fail_if_method(python_callable)
+        super().__init__(*args, **kwargs)
+        self.python_callable = python_callable
+        self.multiple_outputs = multiple_outputs
+        self._kwargs = kwargs
+        self._op_args: List[Any] = []
+        self._called = False
+        self._op_kwargs: Dict[str, Any] = {}
+
+    @staticmethod
+    def _fail_if_method(python_callable):
+        if 'self' in signature(python_callable).parameters.keys():
+            raise AirflowException('@task does not support methods')
+
+    def __call__(self, *args, **kwargs):
+        # If args/kwargs are set, then operator has been called. Raise exception
+        if self._called:
```

Comment:
Curious why the functions can't be reused seems a bit annoying for users, wonder if we can fix this (e.g. late-binding the task_ids or something...) or add a TODO, might be worth adding a comment here. ._checking if it was already called feels a bit hacky as it kind of couples the task execution with global state.

Reply:
Mainly is bc of 1 function == 1 operator. The alternative was the idea that @evgenyshulman proposed which is to use functions as operators generators. This simplifies a bit, but also then you can't use the change the operator later on (it's never accessible in the DAG file itself). I've seen both approaches take here (either 1to1 or 1tomany). For me I like better the 1to1, but mainly bc it allows you to use it as you would use any operator later and feels a bit more intuitive.

Reply:
Do I correctly understand that this will not work?
``` python
@task 
def update_user(user_id: str):
    ...

with DAG(...):
    # Fetch list of users 
    ...
    # Execute task for each user
    for user_id in users_list:
        update_user(user_id)
```

Reply:
Hm, I have mixed feelings:
- `copy` is not self-explanatory in this case imho
- in such case, shouldn't we generate auto id? Or at least can we try to do `update_user(user_id, task_id=user_id)`

Reply:
Main worry is that then what is `update_user`. What you are describing here is using `update_user` as an operator factory. It has it's value, but it also feels too magic to me atm. If `update_user` is a factory, then you can't change the operator instance at all or use it to set non-data dependencies. 

We could capture task_id kwarg and generate a new operator, but then what is `update_user` the first operator, the latest one? What does `update_user` represent?

Reply:
> We could capture task_id kwarg and generate a new operator, but then what is `update_user` the first operator, the latest one? What does `update_user` represent?

For me `update_user` is a function and as a function in can be called many times with different input thus yielding different results. I have never meet "function as a singleton" pattern. If we don't want to generate `task_id` for users then we may consider raising an exception on second invocation when no custom `task_id` is passed. 

My point is: this is a function, I expect to be able to call it as many time as I wish. I expect Airflow to treat each call of this function (in proper context) to be as creating a new task.

> You can either do (1) `update_user(i) for i in range(20)` or (2) `update_user >> other_operation`, but not both. I prefer to support 2nd option as it adapts more to what Airflow already does with operators.

Why should I not be able to do this? This is something that I saw many times. 
```python
first_task = BashOperator()
last_task = BashOperator()

for user_id in users_list:
   first_task >> update_user(user_id) >> last_task
```

Reply:
Yeah, this use case is exactly what we want to support.

The issue being we'd have to generate a new task for each one (doable, that isn't a problem, we just need to work out _what_ task_id to call it).

Perhaps taking a leaf out of pytest or https://pypi.org/project/parameterized/ for how they name functions/test cases?